include/diagnostics/ess_updater_autocovariance.hpp

```diff
@@ -0,0 +1,70 @@
+// VolEsti (volume computation and sampling library)
+// Copyright (c) 2021 Vissarion Fisikopoulos
+// Copyright (c) 2021 Apostolos Chalkis
+
+// Licensed under GNU LGPL.3, see LICENCE file
+
+
+#ifndef DIAGNOSTICS_ESS_UPDATER_AUTOCOVARIANCE_HPP
+#define DIAGNOSTICS_ESS_UPDATER_AUTOCOVARIANCE_HPP
+
+#include <boost/accumulators/accumulators.hpp>
+#include <boost/accumulators/statistics/stats.hpp>
+#include <boost/accumulators/statistics/variance.hpp>
+#include <unsupported/Eigen/FFT>
+#include <complex>
+#include <cmath>
+
+
+/**
+    Compute autocovariance estimates for every lag for the input sequence
+    using the Geyer's stable estimator given in 
+    Charles J. Geyer, Practical Markov Chain Monte Carlo, Statistical Science 1992.
+
+    * @tparam NT number type
+    * @tparam VT vector type
+    * @param samples the sequence of correlated samples
+    * @param auto_cov the autocovariance estimates
+*/
+template <typename NT, typename VT>
+void compute_autocovariance(VT const& samples, VT &auto_cov) 
+{
+    typedef Eigen::FFT<NT> EigenFFT;
+    typedef Eigen::Matrix<std::complex<NT>, Eigen::Dynamic, 1> CVT;
+    EigenFFT fft;
+
+    unsigned int N = samples.size();
+    NT samples_mean = samples.mean();
+    auto_cov.setZero(N);
+
+    // compute normalized samples
+    VT normalized_samples(2 * N);
+    normalized_samples.setZero();
+    normalized_samples.head(N) = samples.array() - samples_mean;
+
+    NT variance = (normalized_samples.cwiseProduct(normalized_samples)).sum();
+    variance *= (1.0 / N);
+    variance += NT(1e-16)*(samples_mean*samples_mean);
```

Comment:
what is the purpose of this computation? why `1e-16`? Is this some magic number that simply works?

Reply:
Yes, it was added to avoid numerical overflow in case where `(normalized_samples.cwiseProduct(normalized_samples)).sum()` is zero.

Now I use `const NT eps = 1e-16;`